docs/rfcs/optional-config-type.md

```diff
@@ -0,0 +1,184 @@
+# Optional[T] type for use in configuration structs
+
+## Overview
+
+In Go, types are by default set to a "zero-value", a supported value for the
+respective type that is semantically equivalent or similar to "empty", but which
+is also a valid value for the type. For many config fields, the zero value is
+not a valid configuration value and can be taken to mean that the option is
+disabled, but in certain cases it can indicate a default value, necessitating a
+way to represent the presence of a value without using a valid value for the
+type.
+
+Using standard Go without inventing any new types, the two most straightforward
+ways to accomplish this are:
+
+1. Using a separate boolean field to indicate whether the field is enabled or
+   disabled.
+2. Making the type a pointer, which makes a `nil` pointer represent that a
+   value is not present, and a valid pointer represent the desired value.
+
+Each of these approaches has deficiencies: Using a separate boolean field
+requires the user to set the boolean field to `true` in addition to setting the
+actual config option, leading to suboptimal UX. Using a pointer value has a few
+drawbacks:
+
+1. It may not be immediately obvious to a new user that a pointer type indicates
+   a field is optional.
+2. The distinction between values that are conventionally pointers (e.g. gRPC
+   configs) and optional values is lost.
+3. Setting a default value for a pointer field when decoding will set the field
+   on the resulting config struct, and additional logic must be done to unset
+   the default if the user has not specified a value.
+4. The potential for null pointer exceptions is created.
+
+## Optional types
+
+Go does not include any form of Optional type in the standard library, but other
+popular languages like Rust and Java do. We could implement something similar in
+our config packages that allows us to address the downsides of using pointers
+for optional config fields.
+
+## Basic definition
+
+A production-grade implementation will not be discussed or shown here, but the
+basic implementation for an Optional type in Go could look something like:
+
+```golang
+type Optional[T any] struct {
+	hasValue bool
+	value    T
+
+	defaultVal T
+}
+
+func Some[T any](value T) Optional[T] {
+	return Optional[T]{value: value, hasValue: true}
+}
+
+func None[T any](value T) Optional[T] {
+	return Optional[T]{}
+}
+
+func WithDefault[T any](defaultVal T) Optional[T] {
```

Comment:
I think `WithDefault` should accept a lambda, `func() T`, not an actual value. The defaults may be runtime sensitive, an example would be the `localhost` vs. `0.0.0.0` that could be controlled by an external mechanism like a feature gate.

Reply:
I had another motivation for function in my prototype https://github.com/open-telemetry/opentelemetry-collector/pull/10260/files#diff-4825c80a1cbc3775a3f3cd1aa7e6d01b6226bc31efaa9ca7b6a1a466b31512f2R36
```
// The reason we pass a function instead of T directly
// is to allow this to work with T being a pointer,
// because we wouldn't want to copy the value of a pointer
// since it might reuse (and override) some shared state.
```

Reply:
Sorry for the confusion here, the reason I didn't have `WithDefault` accept a function was to simplify the implementation to focus on high-level concepts instead of the design of the API. This is not at all supposed to be a carefully-designed or production-ready implementation, which I try to call out above the code block.

If there's a consideration in the design of `WithDefault` that we should be considering for whether the type is worth adding at all, or if using a value instead of a function makes things less clear, I can adjust the example implementation in this document.

Reply:
fair enough, but the "simplified" implementation is different from the one with lambda in about 10 characters.